### ISSUE TITLE
registerInstance can Cause Segmentation Faults

### ISSUE THREAD

**Issue by bluscr33n** [libavg/libavg#683]
The `registerInstance` method can cause segmentation faults when used *wrong*. Although the method takes the child as well as the parent as parameters, it is, obviously, important to call the method on the actual child, which should get registered by its parent.

If the method is not called on the child node, it does not seem to matter at first - the node is displayed as expected. However, if you want to work with the node, segmentation faults can occur.

See the following example:
```
from libavg import app, avg, gesture
from libavg.app import settings


class MainDiv(app.MainDiv):
    def onInit(self):
        container = avg.DivNode(
            size=self.size)
        container.registerInstance(container, self)

        self.node = avg.RectNode(
            size=(200, 200),
            fillopacity=1,
            parent=container)

        self._drag_recognizer = gesture.DragRecognizer(
            eventNode=self.node,
            detectedHandler=self._on_drag_detected,
            moveHandler=self._on_drag_move,
            friction=-1)
        self._start_pos = None

    def _on_drag_detected(self):
        self._start_pos = self.node.pos

    def _on_drag_move(self, offset):
        self.node.pos = self._start_pos + offset


app.App().run(MainDiv(), app_resolution='1400, 800')
```
This example works quite fine - you can drag the white rectangle around. If you change the line
```
container.registerInstance(container, self)
```
to
```
self.registerInstance(container, self)
```
the application crashes as soon as you start dragging.

Another way to crash the application is, in my opinion, more critical. Even if you registered the `container` node right, the whole thing can still go south if you add a second, unrelated container, which was not registered properly. For instance, you can change the `onInit` method from above to the following:
```
container = avg.DivNode(
    size=self.size)
container.registerInstance(container, self)

self.node = avg.RectNode(...)

self._drag_recognizer = gesture.DragRecognizer(...)
self._start_pos = None

container_2 = avg.DivNode(
    size=(self.size.x / 2, self.size.y),
    pos=(self.size.x / 2, 0))
self.registerInstance(container_2, self)
```
In this case, the applications also crashs on drag.

**Comment by scotty007** [comment]
`registerInstance()` is supposed to be used in subclasses of `avg.*Node` only (see [Subclassing](https://www.libavg.de/site/projects/libavg/wiki/Subclassing)). There's no need to use it when just playing with `avg.*Node`s.
If you want to make a node a child of a `DivNode` (the only node with children), use `DivNode.appendChild()` or `DivNode.insertChild*()` (see [Reference](https://www.libavg.de/reference/svn/areanodes.html#libavg.avg.DivNode.appendChild)), or just provide the `DivNode` as `parent` parameter to the child node's `__init()__` call.

**Comment by bluscr33n** [comment]
I'm aware that `registerInstance` is only necessary when working with subclasses - the given example should illustrate how to force the segmentation fault.

In fact, I have faced the issue while working with subclasses. Basically, I initialized a node instance *A* (which inherits from `avg.DivNode`) without directly registering it by a parent (the parent is unknown at this point). Later, the `registerInstance` method is called inside another node class *B*.
At first, I called the method on `self` inside class *B* instead on the instance of *A* - which caused a segmentation fault later.

Once you know about this issue, it is easy to avoid it. However, I think it should be at least mentioned in the documentation.
